numcodecs/shuffle.pyx

```diff
@@ -0,0 +1,99 @@
+# cython: embedsignature=True
+# cython: profile=False
+# cython: linetrace=False
+# cython: binding=False
+# cython: language_level=3
+
+import numpy as np
+from .compat import ensure_bytes, ensure_contiguous_ndarray, ndarray_copy, ensure_ndarray
+from .abc import Codec
+
+
+cdef _doShuffle(const unsigned char[:] src, unsigned char[:] des, int element_size):
+    cdef int count, i, j, offset, byte_index
+    count = len(src) // element_size
```

Comment:
I did a cythonize-annotate, and *this* line has extra code because element_size could be zero and raise an exception; except that we know that it cannot the way we have called it.

The rest I have optimized to
```cython
cimport cython

@cython.boundscheck(False)
@cython.wraparound(False)
cdef void _doShuffle(const unsigned char[::1] src, unsigned char[::1] des, int element_size) nogil:
    cdef int count, i, j, offset, byte_index
    count = len(src) // element_size
    for i in range(count):
        offset = i*element_size
        for byte_index in range(element_size):
            j = byte_index*count + i
            des[j] = src[offset + byte_index]


@cython.boundscheck(False)
@cython.wraparound(False)
cdef void _doUnshuffle(const unsigned char[::1] src, unsigned char[::1] des, int element_size) nogil:
    cdef int count, i, j, offset, byte_index
    count = len(src) // element_size
    for i in range(element_size):
        offset = i*count
        for byte_index in range(count):
            j = byte_index*element_size + i
            des[j] = src[offset+byte_index]
```

Tests pass and are much faster. Should we have a case where we know what the output should be, rather than just checking round-trips?

Reply:
Thanks Martin, incorporated with typing suggestions from John